src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7347,15 +7347,22 @@ private StatementSyntax ParseEmbeddedStatement(bool complexCheck)
                 statement = SyntaxFactory.EmptyStatement(EatToken(SyntaxKind.SemicolonToken));
             }
 
-            // An "embedded" statement is simply a statement that is not a labelled
-            // statement or a declaration statement.
             switch (statement.Kind)
             {
+                // An "embedded" statement is simply a statement that is not a labelled
+                // statement or a declaration statement.
                 case SyntaxKind.LabeledStatement:
                 case SyntaxKind.LocalDeclarationStatement:
                 case SyntaxKind.LocalFunctionStatement:
                     statement = this.AddError(statement, ErrorCode.ERR_BadEmbeddedStmt);
                     break;
+                // Expression statements as embedded statements should be followed by semicolon.
+                case SyntaxKind.ExpressionStatement:
+                    if (((ExpressionStatementSyntax)statement).SemicolonToken.IsMissing)
+                    {
+                        statement = this.AddError(statement, ErrorCode.ERR_SemicolonExpected);
```

Comment:
>statement = this.AddError(statement, ErrorCode.ERR_SemicolonExpected); [](start = 24, length = 70)

We should be testing location of the error. I would expect the error to be attached to the SemicolonToken. Also, it should be done conditionally and the condition should match the one in ```private ExpressionStatementSyntax ParseExpressionStatement(ExpressionSyntax expression)```.